RFC.md

```diff
@@ -0,0 +1,466 @@
+# [RFC] Fused Models
+
+**TLDR**
+- Fused Models are two+ pre-trained models joined together and further tuned to work as one model. This is the approach used for most SOTA Multimodal models currently.
+- This RFC proposes special modules and utils similar to our PEFT implementation to allow for easy manipulation of these fused model parameters and their state dicts.
+
+**WHY**: To support a wide variety of multimodal models and multimodal recipes with a consistent single model API.
+## Context
+The majority of SOTA multimodal models coming out these days are not trained end to end as uniform models, but instead a separate model is trained for each modality and then these models are fused together. In the simplest case, you have an LLM trained on text and an image encoder such as CLIP. You then need to combine these pre-trained models and finetune them so that the LLM can understand the embeddings from the image encoder.
+
+There are many different ways to combine them, and a detailed breakdown of these approaches can be found [here](https://arxiv.org/pdf/2405.17927). At a high level, there are two overarching architectures that most models can be grouped into. **EarlyFusion** and **DeepFusion**. **EarlyFusion** defines a model where a mapping is learned from encoder outputs to token embeddings. Then encoded inputs are converted to tokens and fed into the LLM the same as text tokens. On the other hand, in **DeepFusion**, the output of the encoder is injected directly into the intermediate layers of the LLM, typically with cross attention, following the same approach for encoder decoder language models.
```

Comment:
> **EarlyFusion** defines a model where a mapping is learned from encoder outputs to token embeddings

I don't think I understood this part?

Reply:
For EarlyFusion, the pre-trained encoder model outputs are passed through another model (sometimes just a single layer) before being fed into the LLM as token embeddings. The additional learned parameters between the encoder and decoder learn to convert the encoder embeddings to be in the same latent space as the token embeddings.

Reply:
Does the "other model" here refer to a tokenizer? Or what's an example. Sorry if I'm being dense

Reply:
In the simplest case it's a few extra layers on top of the encoder. I think what's confusing here is that eventually, all of this gets called the encoder. But the way the encoder is built is a pre-trained-encoder + extra-learnable-parameters. These extra parameters are called "fusion_params" in this RFC. For concrete examples, if you look at Llava, the pre-trained encoder is CLIP and the fusion_params is a single linear layer on top. If you look at BLIP2, the pre-trained-encoder is also CLIP and the fusion_params, (or other model) is the QFormer.